sql/analyzer/resolve_columns.go

```diff
@@ -374,83 +373,102 @@ func qualifyExpression(e sql.Expression, symbols availableNames) (sql.Expression
-			// We override the outer scope and qualify this column
-			// only if the current scope provides a definition.
-			levels = levels[len(symbols)-1:]
-		}
-		for _, level := range levels {
-
-			tablesForColumn := symbols.tablesForColumnAtLevel(name, level)
-
-			// If the table exists but it's not available for this node it
-			// means some work is still needed, so just return the column
-			// and let it be resolved in the next pass.
-			// TODO:
-			// if !stringContains(tablesForColumn, table) {
-			// 	return col, nil
-			// }
-
+		sort.Sort(sort.Reverse(sort.IntSlice(levels)))
+		name := strings.ToLower(col.Name())
+		for _, scopeLevel := range levels {
+			tablesForColumn := symbols.tablesForColumnAtLevel(name, scopeLevel)
 			switch len(tablesForColumn) {
 			case 0:
 				// This column could be in an outer scope, keep going
 				continue
 			case 1:
 				if tablesForColumn[0] == "" {
-					return col, transform.SameTree, nil
-				}
-				return expression.NewUnresolvedQualifiedColumn(
-					tablesForColumn[0],
-					col.Name(),
-				), transform.NewTree, nil
-			default:
-				if len(symbols[level].lastRel) > 0 {
+					// This indicates we found a match with an alias definition
+					if canAccessAliasAtCurrentScope || scopeLevel != levels[0] {
+						return expression.NewAliasReference(col.Name()), transform.NewTree, nil
+					}
+				} else {
 					return expression.NewUnresolvedQualifiedColumn(
-						symbols[level].lastRel,
+						tablesForColumn[0],
 						col.Name(),
 					), transform.NewTree, nil
 				}
+			default:
+				aliasesFound := make([]string, 0, len(tablesForColumn))
+				tablesFound := make([]string, 0, len(tablesForColumn))
+				for _, tableName := range tablesForColumn {
+					if tableName == "" {
+						// Regardless of the number of aliases defined with a specific alias name, tablesForColumnAtLevel
+						// currently returns a single empty string to represent them, so check in another datastructure
+						// to see how many alias definitions actually used this name.
+						for range symbols[scopeLevel].availableAliases[name] {
+							aliasesFound = append(aliasesFound, tableName)
+						}
+					} else {
+						tablesFound = append(tablesFound, tableName)
+					}
+				}
+
+				switch node.(type) {
+				case *plan.Sort, *plan.Having:
+					// For order by and having clauses... prefer an alias over a column when there is ambiguity
+					if len(aliasesFound) == 0 {
+						if len(tablesFound) == 1 {
+							return expression.NewUnresolvedQualifiedColumn(tablesFound[0], col.Name()), transform.NewTree, nil
+						} else if len(tablesFound) > 1 {
+							return col, transform.SameTree, sql.ErrAmbiguousColumnOrAliasName.New(col.Name())
+						}
+					} else if len(aliasesFound) == 1 {
+						return expression.NewAliasReference(col.Name()), transform.NewTree, nil
+					} else if len(aliasesFound) > 1 {
+						return col, transform.SameTree, sql.ErrAmbiguousColumnOrAliasName.New(col.Name())
```

Comment:
Yeah... you're correct that duplicate named aliases in the same scope were getting combined. To match MySQL's behavior we need to be better about keeping track of those duplicates. I started moving towards that by changing `scopeLevelSymbols.availableAliases` to be a map of `string` to `[]expression.Alias` so that we do now keep duplicates separate, but `tablesAtLevel()` still returns a single `""` even if there are multiple aliases. The code will check `scopLevelSymbols.availableAliases` for the real alias count and update `aliasesFound`, so that's how this code block would run. 

I still want to take a smaller pass or two on this code in future PRs to keep tidying up a little bit more (e.g. updating `tablesAtLevel()`, fixing some broken queries in the skipped tests), but wanted to get these meatiest changes locked in first.